### ISSUE TITLE
Motivation for custom convolutions

### ISSUE THREAD

**Issue by askorikov** [ahendriksen/msd_pytorch#4]
First of all, thank you for your work on this project! I was wondering, what is the motivation for the custom implementation of convolutions?

**Comment by ahendriksen** [comment]
Hi, 

Great question! 

An efficient implementation of the mixed-scale dense network architecture requires that overlapping buffers can be used as inputs to different convolution operations in the forward pass. In the backward pass, the gradient of all the convolutions with respect to their input must thus be added together. This is difficult to do, because PyTorch prevents overwriting and modifying tensors that are used in a gradient computation. 

In earlier versions of PyTorch (0.3.1 and earlier), implementing the mixed-scale dense network even semi-efficiently was impossible without resorting to C++. It was necessary to make a lot of memory copies to satisfy the memory safety requirements of PyTorch, which some people [did](https://github.com/lvdmaaten/efficient_densenet_pytorch/blob/master/models/densenet_efficient.py).

With the introduction of `torch.nn.grad.conv2d_weight` and `torch.nn.grad.conv2d_input`, it became possible to implement the backward pass using PyTorch functions with reasonable efficiency. This is still not as efficient as using custom convolutions though. For a single 2000x2000 pixel image, the backward pass of msd_pytorch requires 5.8GB GPU memory. With pytorch-provided convolutions, this becomes 7.2GB. 

Necessarily, this reduces the feasibility of using the network with large images, one of the primary use cases of the MS-D network. 

In addition, contrary to what I expected, the custom convolutions are (still) faster than the convolutions provided by PyTorch. I have written up some initial results [here](https://blog.allardhendriksen.nl/cwi-ci-group/20201023170553-benchmarking_custom_convolutions_versus_pytorch_convolutions/). 

All in all, maintaining custom convolutions is a major pain because building conda packages takes forever. In this case, however, I believe it's worth it.

**Comment by ahendriksen** [comment]
You can try for yourself. 

I have uploaded gist that implements the MSD_block structure with just the PyTorch-provided convolutions: 

https://gist.github.com/ahendriksen/a3ffa953ae15c5db3eea154bcc6ccbb5

**Comment by askorikov** [comment]
Thanks for the extended answer! I get it now.

**Comment by ahendriksen** [comment]
You are welcome :)
